rust/src/operations/mod.rs

```diff
@@ -0,0 +1,287 @@
+
+    /// Error returned when errors occur in underlying storage instance
+    #[error("Error in underlying storage backend")]
+    Storage {
+        /// Raw internal StorageError
+        #[from]
+        source: StorageError,
+    },
+
+    /// Error returned when errors occur in Arrow
+    #[error("Error handling arrow data")]
+    Arrow {
+        /// Raw internal ArrowError
+        #[from]
+        source: ArrowError,
+    },
+
+    /// Error returned for errors internal to Datafusion
+    #[error("Error in Datafusion execution engine")]
+    DataFusion {
+        /// Raw internal DataFusionError
+        #[from]
+        source: DataFusionError,
+    },
+}
+
+fn to_datafusion_err(e: impl std::error::Error) -> DataFusionError {
+    DataFusionError::Plan(e.to_string())
+}
+
+/// High level interface for executing commands against a DeltaTable
+pub struct DeltaCommands {
+    table: DeltaTable,
+}
+
+impl DeltaCommands {
+    /// load table from uri
+    pub async fn try_from_uri<T>(uri: T) -> DeltaCommandResult<Self>
+    where
+        T: Into<String>,
+    {
+        let table = get_table_from_uri_without_update(uri.into())?;
+        Ok(Self { table })
+    }
+
+    /// Get a reference to the underlying table
+    pub fn table(&self) -> &DeltaTable {
+        &self.table
+    }
+
+    async fn execute(
+        &mut self,
+        operation: DeltaOperation,
+        plan: Arc<dyn ExecutionPlan>,
+    ) -> DeltaCommandResult<()> {
+        let transaction = Arc::new(DeltaTransactionPlan::new(
+            self.table.table_uri.clone(),
+            plan,
+            operation,
+            None,
+        ));
+
+        let session_ctx = SessionContext::new();
+        let task_ctx = session_ctx.task_ctx();
+        let _ = collect(transaction, task_ctx).await?;
+        self.table.update().await?;
+
+        Ok(())
+    }
+
+    /// Create a new Delta table
+    pub async fn create(
+        &mut self,
+        metadata: DeltaTableMetaData,
+        mode: SaveMode,
+    ) -> DeltaCommandResult<()> {
+        let operation = DeltaOperation::Create {
+            mode,
+            metadata: metadata.clone(),
+            location: self.table.table_uri.clone(),
+            // TODO get the protocol from somewhere central
+            protocol: Protocol {
+                min_reader_version: 1,
+                min_writer_version: 2,
+            },
+        };
+        let plan = Arc::new(CreateCommand::try_new(
+            &self.table.table_uri,
+            operation.clone(),
+        )?);
+
+        self.execute(operation, plan).await
+    }
+
+    /// Write data to Delta table
+    pub async fn write(
+        &mut self,
+        data: Vec<RecordBatch>,
+        mode: SaveMode,
+        partition_columns: Option<Vec<String>>,
+    ) -> DeltaCommandResult<()> {
+        if data.is_empty() {
+            return Ok(());
+        }
+        let schema = data[0].schema();
+        let current_part = match self.table.update().await {
```

Comment:
I think the update call here is not needed because delta writes relies on optimistic concurrency control. I.e. we can always assume the current in memory state is already the latest version. During the table commit write, we will get a file already exists error if we don't have the latest version. When a conflict happens, we can then retry the write after a table update.

Most delta tables only have a single writer, so the table update before write is usually a no op.